### PR TITLE
Using jsonNoDup from Aeson to avoid duplicated fields

### LINKED ISSUE #175: Validation/Check duplicate variables on request Body
with new release now is possible to validate variable name collision on Query body

https://github.com/bos/aeson/commit/cd4ce0454b842205238c718c93bd2b0ebb7a62a8

### PR DESCRIPTION
**Pull request by gvolpe**
closes #175 

Let me know if I'm on the right track or you had something different in mind. Also, if you have some other preferences in terms of coding style / good practices let me know so I can adapt my code :)

### CODE REVIEW

File: src/Data/Morpheus/Server/ClientRegister.hs [hunk 1]
```diff
@@ -64,7 +64,7 @@ updateClientByID id' updateFunc state = modifyMVar_ state (return . map updateCl
 publishUpdates :: (Eq s) => GQLState IO s -> PubPair s -> IO ()
 publishUpdates state (channels, value) = do
   state' <- readMVar state
-  forM_ state' sendMessage
+  traverse_ sendMessage state'
```
Comment (gvolpe) [inline src/Data/Morpheus/Server/ClientRegister.hs]: Unrelated to this PR. `forM_` is now `for_` which is just a flipped version of `traverse_` and I think the latter is more standard so I took the liberty of changing this but let me know if you don't like this, I can remove it.

File: src/Data/Morpheus/Resolve/Resolve.hs [hunk 2]
```diff
@@ -55,9 +59,13 @@ type RootResCon m event query mutation subscription
      , EncodeCon (PublishStream m event) mutation
      , EncodeSubCon m event subscription)
 
-byteStringIO :: Monad m => (GQLRequest -> m GQLResponse) -> ByteString -> m ByteString
+eitherDecodeNoDup :: L.ByteString -> Either String GQLRequest
+eitherDecodeNoDup bs =
+  encode <$> parseOnly jsonNoDup (S.pack . UTF8.encode $ L.unpack bs) >>= eitherDecode
```
Comment (gvolpe) [inline src/Data/Morpheus/Resolve/Resolve.hs]: Not sure if what I'm doing here is correct. I get a `Data.Aeson.Value` from `parseOnly` which I need to encode to `ByteString` and then decode again into `GQLRequest`. There's probably an easier way to do this?

**Comment by gvolpe** [comment]
The CI build is failing with `ghc-8.2.2`. Not sure what the best way to fix it you would prefer.

**Comment by nalchevanidze** [comment]
In the dependencies for aeson-1.4.4.0:
  th-abstraction-0.2.6.0 from stack configuration does not match >=0.2.8.0 && <0.4  (latest
                           matching version is 0.3.1.0)
  needed due to morpheus-graphql-0.1.1 -> aeson-1.4.4.0

In the dependencies for time-compat-1.9.2.2:
   base-orphans-0.7 from stack configuration does not match >=0.8.1 && <0.9  (latest matching
                     version is 0.8.1)
needed due to morpheus-graphql-0.1.1 -> time-compat-1.9.2.2

checkout  `stack-11.10.yaml` i think adding them in extra deps could help them

**Comment by gvolpe** [comment]
@nalchevanidze whenever you have a chance please have a look. The CI build is passing after that last commit 5 days ago. There's only some weird duplication of CI builds.

File: src/Data/Morpheus/Resolve/Resolve.hs (hunk 2 above)
Comment (nalchevanidze) [reply src/Data/Morpheus/Resolve/Resolve.hs]: may we could use use `fromJSON :: FromJSON a => Value -> Result a`

**Comment by nalchevanidze** [comment]
hi, @gvolpe  sorry for late response. i made a comment  and there is also merge Conflict, could you look on it?

**Comment by gvolpe** [comment]
@nalchevanidze done! :)

**Comment by nalchevanidze** [comment]
One more thing. Could you update changelog and specifications?

**Comment by gvolpe** [comment]
Done @nalchevanidze . Let me know if that's okay or you had something different in mind.

**Comment by nalchevanidze** [comment]
> Done @nalchevanidze . Let me know if that's okay or you had something different in mind.

sorry, i think there is some misunderstanding. every graphql `http`  request looks like  

```json 
{"query":"{\n  user {\n    email,,,,,,,,,\n    address {\n      houseNumber\n    }\n  }\n}\n",
"variables":{"userName":"",...}}
```

i parse `http Request` body it with `Aeson` and generate `GQLRequest`

issue was about. that if we receive http `request` with `JSON` body: 

```json 
{"query":"...",
"variables":{"userName":"a", "userName":"b",...}}
```
or 
```json 
{"query":"...",
"variables":{"userName":"a",...}},
"variables":{"userName":"a",...}}
```
and ...

to throw error about duplicate `JSON` fields.

so it should validate duplicate fields on `JSON` object and not on `GraphQL` query selection.

sorry again, i did not describe it properly that why it was ambigious.

**Comment by gvolpe** [comment]
Not sure what should I change, specifications and changelog? What would you like me to add?

**Comment by nalchevanidze** [comment]
Changelog only

**Comment by nalchevanidze** [comment]
The change only affects variables not selection fields.

**Comment by nalchevanidze** [comment]
Done

**Pull request merged**